app/questionnaire/questionnaire_schema.py

```diff
@@ -102,21 +105,53 @@ def when_rules_section_dependencies_by_section(
     ) -> ImmutableDict[str, set[str]]:
         return ImmutableDict(self._when_rules_section_dependencies_by_section)
 
+    @cached_property
+    def when_rules_section_dependencies_for_progress(
+        self,
+    ) -> defaultdict[str, set[str]]:
+        """
+        This method flips the dependencies that were captured for progress value sources so that they can be
+        evaluated properly for when rules, this is because for when rules we need to check for dependencies in
+        previous sections, whereas for progress we are checking for dependent blocks/sections in "future" sections
+        """
+        when_rules_section_dependencies_for_progress = defaultdict(set)
+        for (
+            section,
+            dependent,
+        ) in (
+            self._when_rules_block_dependencies_by_section_for_progress_value_source.items()
+        ):
+            section_dependents = get_flattened_mapping_values(dependent)
+            for dependent_section in section_dependents:
+                when_rules_section_dependencies_for_progress[dependent_section].add(
+                    section
+                )
+
+        for (
+            section,
+            dependents,
+        ) in (
+            self._when_rules_section_dependencies_by_section_for_progress_value_source.items()
+        ):
+            for dependent_section in dependents:
+                when_rules_section_dependencies_for_progress[dependent_section].add(
+                    section
+                )
+        return when_rules_section_dependencies_for_progress
+
     @cached_property
     def when_rules_section_dependencies_by_section_for_progress_value_source(
         self,
-    ) -> ImmutableDict[str, OrderedSet[str]]:
-        return ImmutableDict(
+    ) -> defaultdict[str, OrderedSet[str]]:
+        return (
             self._when_rules_section_dependencies_by_section_for_progress_value_source
         )
 
     @cached_property
```

Comment:
again don't think this is necessary, `self._when_rules_block_dependencies_by_section_for_progress_value_source` could just be `self.when_rules_block_dependencies_by_section_for_progress_value_source` on initialisation

Reply:
Same as the above, we don't modify these values, only fetching it in questionnaire store updater so that's why we are cacheing it here.